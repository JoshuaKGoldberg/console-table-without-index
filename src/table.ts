import logger from "not-a-log";

export type TableParameters = Parameters<(typeof logger)["table"]>;

export function table(...parameters: TableParameters): string {
	const original = logger.table(...parameters);

	// Tables should all start with roughly:
	// ┌─────────┬──────
	// │ (index) │
	// ├─────────┼
	const columnWidth = original.indexOf("┬─");

	const trimmed = original
		.split("\n")
		.map((line) => line.slice(columnWidth))
		.join("\n");

	return "┌" + trimmed.slice(1);
}
