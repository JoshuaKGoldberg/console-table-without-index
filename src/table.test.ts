import { describe, expect, it, vi } from "vitest";

import { table } from "./table.js";

// https://github.com/vitest-dev/vitest/issues/6115
// This is a direct copy & paste of node_modules/not-a-log/not-a-log.js
/*! not-a-logger. MIT License. Jimmy Wärting <https://jimmy.warting.se/opensource> */
vi.mock("not-a-log", async () => {
	const { Console } = await import("console");
	const { Transform } = await import("stream");

	const ts = new Transform({
		transform: (chunk, _, cb) => {
			cb(null, chunk);
		},
	});
	const logger = new Console({ colorMode: false, stderr: ts, stdout: ts });
	const handler = {
		/* eslint-disable @typescript-eslint/ban-ts-comment, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-return */
		// @ts-ignore
		apply(target, _, args) {
			target.apply(logger, args);
			return (ts.read() || "").toString();
		},
		// @ts-ignore
		get(_, prop) {
			// @ts-ignore
			return new Proxy(logger[prop], handler);
		},
		/* eslint-enable */
	};

	const dump = new Proxy(logger, handler);
	return {
		default: dump,
	};
});

describe("table", () => {
	it.each([
		["🍏", "🍌", "🍒"],
		[["🍏", "🍌", "🍒"]],
		[[["🍏", "🍌", "🍒"]]],
		{ emoji: "🍏", fruit: "apple" },
		[
			{ emoji: "🍏", fruit: "apple" },
			{ emoji: "🍌", fruit: "banana" },
			{ emoji: "🍒", fruit: "cherry" },
		],
		[
			[
				{ emoji: "🍏", fruit: "apple" },
				{ emoji: "🍌", fruit: "banana" },
				{ emoji: "🍒", fruit: "cherry" },
			],
		],
	])("%j", async (input) => {
		expect({
			original: "\n" + (await import("not-a-log")).default.table(input),
			result: "\n" + table(input),
		}).toMatchSnapshot();
	});
});
