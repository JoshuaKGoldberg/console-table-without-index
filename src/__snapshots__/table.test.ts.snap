// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`table > "🍏" 1`] = `
{
  "original": "
┌─────────┬────────┐
│ (index) │ Values │
├─────────┼────────┤
│ 0       │ '🍏'   │
│ 1       │ '🍌'   │
│ 2       │ '🍒'   │
└─────────┴────────┘
",
  "result": "
┌────────┐
│ Values │
┼────────┤
│ '🍏'   │
│ '🍌'   │
│ '🍒'   │
┴────────┘
",
}
`;

exports[`table > ["🍏","🍌","🍒"] 1`] = `
{
  "original": "
┌─────────┬──────┬──────┬──────┐
│ (index) │ 0    │ 1    │ 2    │
├─────────┼──────┼──────┼──────┤
│ 0       │ '🍏' │ '🍌' │ '🍒' │
└─────────┴──────┴──────┴──────┘
",
  "result": "
┌──────┬──────┬──────┐
│ 0    │ 1    │ 2    │
┼──────┼──────┼──────┤
│ '🍏' │ '🍌' │ '🍒' │
┴──────┴──────┴──────┘
",
}
`;

exports[`table > [["🍏","🍌","🍒"]] 1`] = `
{
  "original": "
┌─────────┬──────────────────────┐
│ (index) │ 0                    │
├─────────┼──────────────────────┤
│ 0       │ [ '🍏', '🍌', '🍒' ] │
└─────────┴──────────────────────┘
",
  "result": "
┌──────────────────────┐
│ 0                    │
┼──────────────────────┤
│ [ '🍏', '🍌', '🍒' ] │
┴──────────────────────┘
",
}
`;

exports[`table > [{"emoji":"🍏","fruit":"apple"},{"emoji":"🍌","fruit":"banana"},{"emoji":"🍒","fruit":"cherry"}] 1`] = `
{
  "original": "
┌─────────┬─────────────────────────────────┬──────────────────────────────────┬──────────────────────────────────┐
│ (index) │ 0                               │ 1                                │ 2                                │
├─────────┼─────────────────────────────────┼──────────────────────────────────┼──────────────────────────────────┤
│ 0       │ { emoji: '🍏', fruit: 'apple' } │ { emoji: '🍌', fruit: 'banana' } │ { emoji: '🍒', fruit: 'cherry' } │
└─────────┴─────────────────────────────────┴──────────────────────────────────┴──────────────────────────────────┘
",
  "result": "
┌─────────────────────────────────┬──────────────────────────────────┬──────────────────────────────────┐
│ 0                               │ 1                                │ 2                                │
┼─────────────────────────────────┼──────────────────────────────────┼──────────────────────────────────┤
│ { emoji: '🍏', fruit: 'apple' } │ { emoji: '🍌', fruit: 'banana' } │ { emoji: '🍒', fruit: 'cherry' } │
┴─────────────────────────────────┴──────────────────────────────────┴──────────────────────────────────┘
",
}
`;

exports[`table > {"emoji":"🍏","fruit":"apple"} 1`] = `
{
  "original": "
┌─────────┬─────────┐
│ (index) │ Values  │
├─────────┼─────────┤
│ emoji   │ '🍏'    │
│ fruit   │ 'apple' │
└─────────┴─────────┘
",
  "result": "
┌─────────┐
│ Values  │
┼─────────┤
│ '🍏'    │
│ 'apple' │
┴─────────┘
",
}
`;

exports[`table > {"emoji":"🍏","fruit":"apple"} 2`] = `
{
  "original": "
┌─────────┬───────┬──────────┐
│ (index) │ emoji │ fruit    │
├─────────┼───────┼──────────┤
│ 0       │ '🍏'  │ 'apple'  │
│ 1       │ '🍌'  │ 'banana' │
│ 2       │ '🍒'  │ 'cherry' │
└─────────┴───────┴──────────┘
",
  "result": "
┌───────┬──────────┐
│ emoji │ fruit    │
┼───────┼──────────┤
│ '🍏'  │ 'apple'  │
│ '🍌'  │ 'banana' │
│ '🍒'  │ 'cherry' │
┴───────┴──────────┘
",
}
`;
